File: flutter-dart/projects/calculator.dart
// PROJECT: Pure Dart CLI Calculator | dart calculator.dart

import 'dart:io';

// TODO 1 (Mini): stdin REPL, tokenize "3 + 4", compute and print
//   void main() {
//     while (true) {
//       stdout.write(">> ");
//       final line = stdin.readLineSync();
//       if (line == null || line == "exit") break;
//       try {
//         final tokens = line.trim().split(RegExp(r'\s+'));
//         final a = double.parse(tokens[0]);
//         final op = tokens[1];
//         final b = double.parse(tokens[2]);
//         final result = switch (op) {
//           "+" => a + b,
//           "-" => a - b,
//           "*" => a * b,
//           "/" => b == 0 ? throw Exception("Division by zero") : a / b,
//           "%" => a % b,
//           _ => throw Exception("Unknown operator: $op"),
//         };
//         print(result);
//       } catch (e) {
//         print("Error: $e");
//       }
//     }
//   }

// TODO 2 (Intermediate): Tokenizer + recursive descent parser
//   - Use RegExp to scan: r'\d+\.?\d*|[+\-*\/()%]'
//   - Token types: Num, Plus, Minus, Star, Slash, Percent, LParen, RParen
//   - Grammar:
//       expr   → term (('+' | '-') term)*
//       term   → factor (('*' | '/') factor)*
//       factor → '-' factor | NUMBER | '(' expr ')'

// TODO 3 (Advanced): Variable assignment
//   - Parse "x = 5 + 3" → store in Map<String, double>
//   - Allow referencing variables in subsequent expressions

void main() {
  print("TODO: implement Dart calculator");
}

File: js-ts/projects/calculator.ts
// =============================================================================
// PROJECT: TypeScript CLI Calculator
// =============================================================================
// TODO 1 (Mini): CLI calculator — node calculator.js "3 + 4 * 2"
//   - Tokenize and parse expression
//   - Respect operator precedence (PEMDAS)
//   - Handle division by zero
//
// TODO 2 (Intermediate): Interactive REPL
//   - Variable assignment: x = 5 + 3
//   - Built-ins: sqrt(16), abs(-5), floor(3.7)
//   - Commands: history, vars, quit
//
// TODO 3 (Advanced): Typed AST
//   type Expr = NumberLit | BinaryOp | UnaryOp | Variable | FunctionCall
//   parse(tokens:Token[]): Expr
//   evaluate(expr:Expr, env:Map<string,number>): number
//
// Run: npx ts-node calculator.ts
// =============================================================================

type TokenType = "number" | "op" | "lparen" | "rparen" | "ident";
interface Token { type: TokenType; value: string; }

function tokenize(input: string): Token[] {
  // TODO: scan input char-by-char, emit tokens
  return [];
}

function main() {
  const rl = require("readline").createInterface({ input: process.stdin, output: process.stdout });
  console.log("TypeScript Calculator — type 'quit' to exit");
  const ask = () => {
    rl.question("> ", (line: string) => {
      if (line.trim() === "quit") { rl.close(); return; }
      // TODO: evaluate and print
      console.log("(not implemented)");
      ask();
    });
  };
  ask();
}

main();
